Remake Puppet introduction (#1039)
theforeman/foreman-documentation: guides/common/modules/con_puppet-architecture.adoc

[id="Puppet_architecture_{context}"]
= Puppet Architecture

> Puppet is built on client-server architecture.

Reviewer: To me, this makes less sense than the original. The former tells me how Puppet operates while I cannot really understand why I am reading the suggested change.

Author: I was trying to elaborate on the internal workings of Puppet and how it's integrated with {Project} -- it seemed insufficient for a new user and a bit inaccurate before (based on feedback from engineers).
For example, it didn't seem clear to me from the original text if the catalog is the same as the define state of a host. Further, the ENC seemed like an important term and as such should be clarified how it's related to the overall Puppet flow (client-server communication).

Reviewer: I was speaking for line 4 alone!

Author: Oh. Okay, reverting.

The Puppet server is the central component, where the configuration definitions live.
{ProjectServer} or any {SmartProxies} are typically deployed with the Puppet server and {Project} acts as an https://puppet.com/docs/puppet/7/nodes_external.html[External Node Classifier (ENC)] for such Puppet server.
Managed hosts run the Puppet agent as a client that communicates with the Puppet server.

The Puppet agent collects _facts_ about a host and reports them to the Puppet server on each run.
You can display the Puppet _facts_ in JSON format by running `puppet facts` on a host.

The Puppet server forwards _facts_ to {Project} and {Project} stores them for later use.
Based on the _facts_ and other definitions, {Project} constructs the ENC answer to the Puppet server.

> The Puppet server compiles a catalog based on the ENC answer and sends the catalog to the Puppet agent.

Author: I was just wondering about the same!

The Puppet agent evaluates the system state on the host.
If the Puppet agent finds differences (_drifts_) between the defined state in the catalog and the actual state,
it enforces correction of the host's state.
The Puppet agent then reports correction results back to the Puppet server, which reports them to {Project}.

> The system state of a host is defined by means of the catalog.

Author: Maybe this sentence is actually redundant.

> The system state of a host is defined by means of the catalog.

Author: I was trying here to link this paragraph to the previous mention of the catalog and then link it to Puppet modules.

> A catalog is compiled from Puppet manifests of one or more Puppet modules assigned to the host.

Author: See above

> A Puppet module is a collection of classes, manifests, resources, files, and templates.

Reviewer: ```suggestion
Puppet module:: 
A collection of classes, manifests, resources, files, and templates.
```
A definition list for all the new terms would probably be easier on the eye and easier for scanning.

Author: @melcorr Would this work?

The Puppet modules work as components of host configuration definitions.

> A Puppet module can be parametrized using Smart Class parameters if the module supports the use of parameters.

Reviewer: ```suggestion
You can add custom parameters to a Puppet module using Smart Class parameters if the module supports the use of parameters.
```
Avoid the passive if possible.
